File: src/ios/ChannelIOPlugin.m
#import "ChannelIOPlugin.h"
#import <ChannelIO/ChannelIO-swift.h>


@implementation ChannelIOPlugin

- (void)pluginInitialize {
    NSLog(@"Starting Channel IO plugin");
}

- (void)startWithGuestUser:(CDVInvokedUrlCommand *)command {
    NSString* pluginKey = [command.arguments objectAtIndex:0];

    ChannelPluginSettings *settings = [[ChannelPluginSettings alloc] init];
    [settings setPluginKey:pluginKey];

    [ChannelIO bootWith:settings profile:nil completion:^(ChannelPluginCompletionStatus status, User* user) {
      [ChannelIO trackWithEventName:@"init_cordova_channelio" eventProperty:nil];
    }];

    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)startWithRegisteredUser:(CDVInvokedUrlCommand *)command {
    NSString* pluginKey = [command.arguments objectAtIndex:0];
    NSString* userId = [command.arguments objectAtIndex:1];
    NSString* userName = [command.arguments objectAtIndex:2];

    ChannelPluginSettings *settings = [[ChannelPluginSettings alloc] init];
    [settings setPluginKey:pluginKey];
    [settings setMemberId:userId];

    Profile *profile = [[Profile alloc] init];
    [profile setWithName:userName];

    [ChannelIO bootWith:settings profile:profile completion:^(ChannelPluginCompletionStatus status, User* user) {
      [ChannelIO trackWithEventName:@"init_cordova_channelio" eventProperty:nil];
    }];

    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)startWithRegisteredUserAndEmail:(CDVInvokedUrlCommand *)command {
    NSString* pluginKey = [command.arguments objectAtIndex:0];
    NSString* userId = [command.arguments objectAtIndex:1];
    NSString* userName = [command.arguments objectAtIndex:2];
    NSString* userEmail = [command.arguments objectAtIndex:3];

    ChannelPluginSettings *settings = [[ChannelPluginSettings alloc] init];
    [settings setPluginKey:pluginKey];
    [settings setMemberId:userId];

    Profile *profile = [[Profile alloc] init];
    [profile setWithName:userName];
    [profile setWithEmail:userEmail];

    [ChannelIO bootWith:settings profile:profile completion:^(ChannelPluginCompletionStatus status, User* user) {
      [ChannelIO trackWithEventName:@"init_cordova_channelio" eventProperty:nil];
    }];

    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)setDeviceToken:(CDVInvokedUrlCommand *)command {
    NSString* token = [command.arguments objectAtIndex:0];
    if ([token isEqual:[NSNull null]]){
      NSLog(@"token is null");
    }else{
      [ChannelIO initPushTokenWithTokenString:token];
    }

    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)openLauncher:(CDVInvokedUrlCommand *)command {
    [ChannelIO openWithAnimated:YES];

    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)showLauncher:(CDVInvokedUrlCommand *)command {
    [ChannelIO showWithAnimated:YES];

    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)hideLauncher:(CDVInvokedUrlCommand *)command {
    [ChannelIO hideWithAnimated:YES];

    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)track:(CDVInvokedUrlCommand *)command {
    NSString* eventName = [command.arguments objectAtIndex:0];
    NSDictionary<NSString *, id>* eventProperty = [command.arguments objectAtIndex:1];

    [ChannelIO trackWithEventName:eventName eventProperty:eventProperty];

    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

@end

File: src/android/ChannelIOPlugin.java
package by.eeenbnb.cordova.channelio;

import by.chemerisuk.cordova.support.CordovaMethod;
import by.chemerisuk.cordova.support.ReflectiveCordovaPlugin;

import org.apache.cordova.CallbackContext;

import com.zoyi.channel.plugin.android.ChannelIO;
import com.zoyi.channel.plugin.android.Profile;
import com.zoyi.channel.plugin.android.ChannelPluginSettings;
import com.zoyi.channel.plugin.android.OnBootListener;
import com.zoyi.channel.plugin.android.ChannelPluginCompletionStatus;
import com.zoyi.channel.plugin.android.User;

import java.util.Map;


public class ChannelIOPlugin extends ReflectiveCordovaPlugin {
  private static final String TAG = "ChannelIOPlugin";
  private static ChannelIOPlugin instance;

  @Override
  protected void pluginInitialize() {
    ChannelIOPlugin.instance = this;
    ChannelIO.initialize(this.cordova.getActivity().getApplication());
  }

  @CordovaMethod
  private void startWithGuestUser(String pluginKey, CallbackContext callbackContext) {
    ChannelPluginSettings settings = new ChannelPluginSettings(pluginKey);
    ChannelIO.boot(settings, new InitEvent());
  }

  @CordovaMethod
  private void startWithRegisteredUser(String pluginKey, String userId, String name, CallbackContext callbackContext) {
    ChannelPluginSettings settings = new ChannelPluginSettings(pluginKey);
    settings.setMemberId(userId);

    Profile profile = Profile.create()
       .setName(name);

    ChannelIO.boot(settings, profile, new InitEvent());
  }

  @CordovaMethod
  private void startWithRegisteredUserAndEmail(String pluginKey, String userId, String name, String email, CallbackContext callbackContext) {
    ChannelPluginSettings settings = new ChannelPluginSettings(pluginKey);
    settings.setMemberId(userId);

    Profile profile = Profile.create()
         .setName(name)
         .setEmail(email);

    ChannelIO.boot(settings, profile, new InitEvent());
  }

  @CordovaMethod
  private void setDeviceToken(String token, CallbackContext callbackContext) {
    ChannelIO.initPushToken(token);
  }

  @CordovaMethod
  private void openLauncher(CallbackContext callbackContext) {
    ChannelIO.open(instance.webView.getContext());
  }

  @CordovaMethod
  private void showLauncher(CallbackContext callbackContext) {
    ChannelIO.show();
  }

  @CordovaMethod
  private void hideLauncher(CallbackContext callbackContext) {
    ChannelIO.hide();
  }

  @CordovaMethod
  private void track(String eventName, Map<String, Object> properties, CallbackContext callbackContext) {
    ChannelIO.track(eventName, properties);
  }
}

class InitEvent implements OnBootListener {
  public void onCompletion(ChannelPluginCompletionStatus status, User guest){
    ChannelIO.track("init_cordova_channelio");
  }
}
